File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/4673_셀프넘버/4673.py
MAX_NUMBER = 10000
is_selfnum = [True] * (MAX_NUMBER + 1)

def generate(seed):
    num = seed
    while seed > 0:
        num += seed % 10
        seed //= 10
    return num

for i in range(1, MAX_NUMBER + 1):
    seed = i
    while True:
        num = generate(seed)
        if num > MAX_NUMBER:
            break
        is_selfnum[num] = False
        seed = num

for i in range(1, MAX_NUMBER + 1):
    if is_selfnum[i]:
        print(i)

File: Part1_강의자료/Ch04_전체탐색_시뮬레이션/문제별코드/4673_셀프넘버/4673.cpp
#include <stdio.h>
#include <algorithm>
using namespace std;

#define MAX_NUMBER 10000
bool is_selfnum[MAX_NUMBER + 1];

int generate(int seed)
{
    int num = seed;
    while (seed > 0)
    {
        num += seed % 10;
        seed /= 10;
    }
    return num;
}

int main()
{
    fill(is_selfnum, is_selfnum + MAX_NUMBER + 1, true);

    for (int i = 1; i <= MAX_NUMBER; i++)
    {
        int seed = i;
        while (1)
        {
            int num = generate(seed);
            if (num > MAX_NUMBER)
                break;
            is_selfnum[num] = false;
            seed = num;
        }
    }

    for (int i = 1; i <= MAX_NUMBER; i++)
        if (is_selfnum[i])
            printf("%d\n", i);
    return 0;
}
